src/main/java/org/checkerframework/checker/builder/TypesafeBuilderAnnotatedTypeFactory.java

```diff
@@ -107,115 +117,85 @@ public Void visitMethodInvocation(
+      if (!methodTree.getModifiers().getFlags().contains(Modifier.ABSTRACT)) {
+        return false;
       }
+      ClassTree enclosingClass = TreeUtils.enclosingClass(getPath(methodTree));
 
-      if (!AnnotationUtils.hasElementValue(a2, "value")) {
-        return a1;
+      if (enclosingClass == null) {
+        return false;
       }
+      //      System.out.println(node.getName().toString() + " is visited method in class " +
+      // enclosingClass.getSimpleName());
 
-      Set<String> a1Val = new LinkedHashSet<>(getValueOfAnnotationWithStringArgument(a1));
-      Set<String> a2Val = new LinkedHashSet<>(getValueOfAnnotationWithStringArgument(a2));
-      a1Val.addAll(a2Val);
-      return createCalledMethods(a1Val.toArray(new String[0]));
-    }
+      boolean inAutoValueBuilder = hasAnnotation(enclosingClass, AutoValue.Builder.class);
 
-    /**
-     * LUB in this type system is set intersection of the arguments of the two annotations, unless
-     * one of them is bottom, in which case the result is the other annotation.
-     */
-    @Override
-    public AnnotationMirror leastUpperBound(final AnnotationMirror a1, final AnnotationMirror a2) {
-      if (AnnotationUtils.areSame(a1, BOTTOM)) {
-        return a2;
-      } else if (AnnotationUtils.areSame(a2, BOTTOM)) {
-        return a1;
-      }
-
-      if (!AnnotationUtils.hasElementValue(a1, "value")) {
-        return a1;
+      if (inAutoValueBuilder) {
+        Element classElem = TreeUtils.elementFromTree(enclosingClass);
+        Element returnTypeElem = TreeUtils.elementFromTree(methodTree.getReturnType());
+        return classElem.equals(returnTypeElem);
       }
+      return false;
+    }
+  }
 
-      if (!AnnotationUtils.hasElementValue(a2, "value")) {
-        return a2;
-      }
+  private static boolean hasAnnotation(
+      ClassTree enclosingClass, Class<? extends Annotation> annotClass) {
+    return enclosingClass.getModifiers().getAnnotations().stream()
+        .map(TreeUtils::annotationFromAnnotationTree)
+        .anyMatch(anm -> AnnotationUtils.areSameByClass(anm, annotClass));
+  }
 
-      Set<String> a1Val = new LinkedHashSet<>(getValueOfAnnotationWithStringArgument(a1));
-      Set<String> a2Val = new LinkedHashSet<>(getValueOfAnnotationWithStringArgument(a2));
-      a1Val.retainAll(a2Val);
-      return createCalledMethods(a1Val.toArray(new String[0]));
+  private class TypesafeBuilderTypeAnnotator extends TypeAnnotator {
+    public TypesafeBuilderTypeAnnotator(AnnotatedTypeFactory aTypeFactory) {
+      super(aTypeFactory);
     }
 
-    /** isSubtype in this type system is subset */
     @Override
-    public boolean isSubtype(final AnnotationMirror subAnno, final AnnotationMirror superAnno) {
-      if (AnnotationUtils.areSame(subAnno, BOTTOM)) {
-        return true;
-      } else if (AnnotationUtils.areSame(superAnno, BOTTOM)) {
-        return false;
+    public Void visitExecutable(AnnotatedTypeMirror.AnnotatedExecutableType t, Void p) {
```

Comment:
This injects the `@CalledMethods` annotation computed by the `AutoValueBuilderAnnotatedTypeFactory`